### DISCUSSION TITLE
How to reduce the number of fonts during startup

### DISCUSSION THREAD

**Discussion by MaBecker** [ImageMagick/ImageMagick#4423] [Help] [answered]
Hi,

beaglebone  with  4.4.155-ti-r155 and Version: ImageMagick 6.9.7-4 Q16 arm 20170114 http://www.imagemagick.org

rendering takes  about 10sec for this statement:

```bash
convert -background white -fill black -gravity South -pointsize 400 label:"42"  -gravity North -pointsize 100 caption:'This is a text'  -extent 1024x758   -depth 8 gray:xxx.raw
```
Running this with strace shows about handling 750 .afm and 300 .pfb

Is there a way how to reduce the number of fonts to get the output quicker?

**Comment by urban-warrior** [comment]
Use an explicit font, `-font arial.ttf`, for example.  Now try your strace.  Does it still try to load all those fonts?

On our system, your command took 0.2 seconds to complete.

**Comment by MaBecker** [comment]
Hi, thanks for your hint.

Ok, picked `-font Helvetica` because it's available on my system

```bash
root:~# time convert -font Helvetica -background white -fill black -gravity South -pointsize 400 label:"42"  -gravity North -pointsize 100 caption:'This is a text'  -extent 1024x758 -depth 8 gray:xxx.raw

real	0m4.333s
user	0m2.000s
sys	0m0.128s
```

same as without `-font Helvetica`

attached strace as txt:

[convert_helvetica_01.strace.txt](https://github.com/ImageMagick/ImageMagick/files/7446524/convert_helvetica_01.strace.txt)

**Comment by urban-warrior** [comment]
Not quite what we were looking for.  We want an explicit font so no font searching is required.  Find any font with a .ttf extension.  For example, `-font arial.ttf`.  If your command is still slow any performance issues are not related to fonts.

We tried your command on a relatively modest VM with ImageMagick 6.9.10-86 and it ran in half a second.  You might want to try a modern release of ImageMagick for improved performance.

**Comment by MaBecker** [comment]
The beaglebone black is a system with ARM335x 1GHz and  512MB DDR3 RAM 4GB 8-bit eMMC on-board flash storage - so not a racing car.  

Ok, added arial.ttf to font folder and `convert -list font` displays the new entry.

```bash
time convert -font arial.ttf -background white -fill black -gravity South -pointsize 400 label:"42"  -gravity North -pointsize 100 caption:'This is a text'  -extent 1024x758 -depth 8 gray:xxx.raw
convert-im6.q16: unable to read font `arial.ttf' @ warning/annotate.c/RenderType/964.

real	0m2.766s
user	0m1.264s
sys	0m0.100s
```

That's a nice improvement - thanks.

Should I worry about the warning?

**Comment by urban-warrior** [comment]
You need an actual TrueType font file that ends with the `.ttf` extension.  Use Google to find one.  For testing, it can be any .ttf file.  If it works properly, there should be no warning.  Once it does work, post the timing.

**Comment by snibgo** [comment]
> unable to read font `arial.ttf'

Perhaps you need to supply the full path to the TTF file.

**Comment by MaBecker** [comment]
Ok, using `-family Arial` works without warning for the used version. There is no more improvement.

**Comment by urban-warrior** [comment]
Again, that's not we asked.  We need to identify if any perceived performance problem is related to fonts.  To do so, we need a font file to suppress font searching.  That requires a .ttf file on disk and a reference to it with the `-font` option.  The `-family` option is a font search option.

**Comment by MaBecker** [comment]
Sure, but I tried several Arial.ttf files,  but none was working without warning.

**Comment by urban-warrior** [comment]
Try `ls -l arial.ttf`.  Confirm the file is found in the same folder that  you issue the `magick` command.

**Comment by MaBecker** [comment, marked as answer]
Ok, that's the best result  for running this statement without font searching on a beaglebone black rev c:
 
```bash

ls -l /usr/share/fonts/truetype/arial/Arial.ttf
-rwxr-xr-x 1 root root 311636 Oct 30 19:19 /usr/share/fonts/truetype/arial/Arial.ttf

time convert -font /usr/share/fonts/truetype/arial/Arial.ttf -background white -fill black -gravity South -pointsize 400 label:"42"  -gravity North -pointsize 100 caption:'This is a text'  -extent 1024x758 -depth 8 gray:xxx.raw

real	0m2.297s
user	0m1.012s
sys	0m0.116s
```

Many Thanks for your support.
